### DISCUSSION TITLE
Arduino Sleep + setPowerSave ≠ Wake

### DISCUSSION THREAD

**Discussion by glasseffects** [olikraus/u8g2#1571] [Q&A]
Hello everyone, 

I need some help while working with the setPowerSave feature in u8g2 and Arduino low power mode. Basically, the issue seems that the CPU doesn't seem to wake up if power save is enabled in u8g2. Its likely that my implementation is wrong, but I'm not sure how else to implement this feature. The setup is using an Arduino Pro mini 3.3v, connected to 128x64 OLED display via I2C. Please help me out on this:

```
#include <U8g2lib.h> 
#include "LowPower.h"

U8G2_SSD1306_128X64_NONAME_1_HW_I2C u8g2(U8G2_R0, /* reset=*/ U8X8_PIN_NONE);
void setup()
{
  Serial.begin(9600);
  u8g2.begin();
}

void loop()
{
  Serial.println("About to sleep");
  u8g2.setPowerSave(1); 
  delay(1000);

  LowPower.powerDown(SLEEP_2S, ADC_OFF, BOD_OFF);
  Serial.println("Awake");
  u8g2.setPowerSave(0);
  Graphics();
  delay(1000);
}

void Graphics() {
  u8g2.firstPage();
  do {
    u8g2.setFont(u8g2_font_nokiafc22_tu);
    u8g2.setCursor(0, 7);
    u8g2.print("TEST");
  } while ( u8g2.nextPage() );
} 
```

Below is the output when "LowPower.powerDown(SLEEP_2S, ADC_OFF, BOD_OFF);" in the loop section is commented out. 

```
18:50:30.774 -> About to sleep
18:50:31.248 -> Awake
18:50:31.783 -> About to sleep
18:50:32.278 -> Awake
18:50:32.828 -> About to sleep
18:50:33.322 -> Awake
18:50:33.867 -> About to sleep
18:50:34.337 -> Awake
```

Here's the output when "u8g2.begin()" in the setup section is disabled:
```
18:54:43.416 -> About to sleep
18:54:46.130 -> Awake
18:54:46.696 -> About to sleep
18:54:49.418 -> Awake
18:54:49.976 -> About to sleep
18:54:52.683 -> Awake
18:54:53.209 -> About to sleep
18:54:55.920 -> Awake
18:54:56.477 -> About to sleep
18:54:59.211 -> Awake
18:54:59.751 -> About to sleep
18:55:02.487 -> Awake
```

And, here is the output when both are enabled. This is where the issue occurs: 

`18:56:38.881 -> About to sleep
18:56:41.594 -> Awake
`
The cycle does not go beyond this point. Please help me to resolve this issue.
Thanks in advance.

**Comment by olikraus** [comment]
Just to clarify: The setPowerSave feature in u8g2 has nothing todo with the micro controller, it just reduces power for the display.

**Reply by glasseffects** [reply]
Thanks for replying to the post Oliver.

I'm aware of the distinction of the two. I've tried in many ways to make the the arduino sleep function work with power save function in u8g2, but the behavior of the combination is contrary to expectation. Can you please guide how I can make things work in this case.

Thanks.

**Comment by olikraus** [comment]
What I understood is this: You system just stops working at a certain time.

However, I do not think this is related to  u8g2.setPowerSave at all. I think if you would remove both u8g2.setPowerSave statements, then still the system would stop working.

Instead I think the problem is caused by some watchdog timeout. Maybe you need to call the yield() function in the while loop of the Graphics() procedure.

https://www.arduino.cc/en/Reference/SchedulerYield

https://stackoverflow.com/questions/34497758/what-is-the-secret-of-the-arduino-yieldfunction

**Reply by glasseffects** [reply]
Oliver, thank you for the reply. I read the main query and the replies related to the function and implementation of yield() function. However, there are contrasting views upon the result of yield() function on AVR microprocessors. My awareness of computing language is limited, so my replies may not be as you'd expect from someone working with these nifty devices. So, my apologies here for that. 

Meanwhile, I tried to implement it is the following way, unsure of it's systematic use, as a result the issue persistes. Please let me know where I can make necessary changes.
`  } while (u8g2.nextPage());
  yield();`

Thanks!

**Reply by olikraus** [reply]
It should be like this:
``` C+
void Graphics() {
  u8g2.firstPage();
  do {
    u8g2.setFont(u8g2_font_nokiafc22_tu);
    u8g2.setCursor(0, 7);
    u8g2.print("TEST");
    yield();
  } while ( u8g2.nextPage() );
} 
```

**Reply by glasseffects** [reply]
Thanks a lot Oliver for your help. I tried as you suggested, but it seems like there's some routine that is eluding detection. I'll see if I can do the project without sleep function without loosing much upon battery. The entire project consumes about 22mA, and with OLED powersave function saves 11mA from that. So maybe I'll toss in a larger battery :)

Thanks man!

**Reply by glasseffects** [reply]
Just an update. The previous situation exits. However, I decided to add a soft reset function to reset the CPU after a few minutes of inactivity which is called just after the CPU is put to sleep. So the moment the CPU wakes up, its is reset and everything starts afresh.
